optuna/study/_tell.py

```diff
@@ -97,106 +132,70 @@ def _tell_with_warning(
-
-    frozen_trial = study._storage.get_trial(trial_id)
-    warning_message = None
-
+    frozen_trial = _get_frozen_trial(study, trial)
     if frozen_trial.state.is_finished() and skip_if_finished:
         _logger.info(
-            f"Skipped telling trial {trial_number} with values "
+            f"Skipped telling trial {frozen_trial.number} with values "
             f"{values} and state {state} since trial was already finished. "
             f"Finished trial has values {frozen_trial.values} and state {frozen_trial.state}."
         )
         return copy.deepcopy(frozen_trial)
-    elif frozen_trial.state == TrialState.WAITING:
-        raise ValueError("Cannot tell a waiting trial.")
+    elif frozen_trial.state != TrialState.RUNNING:
+        raise ValueError(f"Cannot tell a {frozen_trial.state.name} trial.")
+
+    _check_state_and_values(state, values)
+
+    warning_message = None
+
+    if state == TrialState.COMPLETE:
+        assert values is not None
 
-    if state == TrialState.PRUNED:
+        values_conversion_failure_message = _check_values(study, values)
+        if values_conversion_failure_message is not None:
+            raise ValueError(values_conversion_failure_message)
+    elif state == TrialState.PRUNED:
         # Register the last intermediate value if present as the value of the trial.
         # TODO(hvy): Whether a pruned trials should have an actual value can be discussed.
         assert values is None
 
         last_step = frozen_trial.last_step
         if last_step is not None:
-            values = [frozen_trial.intermediate_values[last_step]]
-
-    values, values_conversion_failure_message = _check_and_convert_to_values(
-        len(study.directions), values, trial_number
-    )
-
-    if state == TrialState.COMPLETE and values_conversion_failure_message is not None:
-        raise ValueError(values_conversion_failure_message)
+            value = frozen_trial.intermediate_values[last_step]
+            # intermediate_values can be unacceptable value, i.e., NaN.
+            if _check_values(study, value) is None:
+                values = [value]
+    elif state is None:
+        if values is None:
+            values_conversion_failure_message = "The value None could not be cast to float."
+        else:
+            values_conversion_failure_message = _check_values(study, values)
 
-    if state is None:
         if values_conversion_failure_message is None:
             state = TrialState.COMPLETE
         else:
             state = TrialState.FAIL
+            values = None
             if not suppress_warning:
                 warnings.warn(values_conversion_failure_message)
             else:
                 warning_message = values_conversion_failure_message
 
     assert state is not None
 
+    if values is not None:
+        if isinstance(values, Sequence):
+            values = [float(value) for value in values]
+        else:
+            values = [float(values)]
+
```

Comment:
Can we simplify `_check_values` a little by doing this conversion first?

Reply:
This conversion is not safe without value checking and requires error handling.  I think it is better to check first and convert later.

Reply:
Oh, sorry for not being clear. I meant the conversion without cast (https://github.com/optuna/optuna/pull/3841/files#diff-b54e13f298eb2cf1237096a8de699cf349e84774b96f8458ef8b8dc0bc03615aR81-R84) can be done first.